airflow/hooks/subprocess.py

```diff
@@ -61,41 +64,68 @@ def run_command(
-                preexec_fn=pre_exec,
-            )
-
-            self.log.info("Output:")
-            line = ""
-            if self.sub_process is None:
-                raise RuntimeError("The subprocess should be created here and is None!")
-            if self.sub_process.stdout is not None:
-                for raw_line in iter(self.sub_process.stdout.readline, b""):
-                    line = raw_line.decode(output_encoding, errors="backslashreplace").rstrip()
-                    self.log.info("%s", line)
-
-            self.sub_process.wait()
-
-            self.log.info("Command exited with return code %s", self.sub_process.returncode)
-            return_code: int = self.sub_process.returncode
+
+        safe_cleanup = False
+
+        try:
+            with contextlib.ExitStack() as stack:
+                if cwd is None:
+                    # TemporaryDirectory will call shutil.rmtree() internally when the context exits. On
+                    # Windows, shutil.rmtree() is unreliable and there is a race condition, where even after
+                    # self.sub_process.wait(), the process will still be holding onto the directory causing
+                    # an exception. The work-around is to call shutil.rmtree() in a retry loop. If we're not
+                    # running under Windows, shutil.rmtree() is reliable and no retry loop is needed.
+
+                    cwd = stack.enter_context(TemporaryDirectory(prefix="airflowtmp"))
+                    safe_cleanup = IS_WINDOWS
+
+                def pre_exec():
+                    # Restore default signal disposition and invoke setsid
+                    for sig in ("SIGPIPE", "SIGXFZ", "SIGXFSZ"):
+                        if hasattr(signal, sig):
+                            signal.signal(getattr(signal, sig), signal.SIG_DFL)
+                    os.setsid()
+
+                self.log.info("Running command: %s", command)
+
+                self.sub_process = Popen(
+                    command,
+                    stdout=PIPE,
+                    stderr=STDOUT,
+                    cwd=cwd,
+                    env=env if env or env == {} else os.environ,
+                    preexec_fn=pre_exec if not IS_WINDOWS else None,
+                )
+
+                self.log.info("Output:")
+                line = ""
+                if self.sub_process is None:
+                    raise RuntimeError("The subprocess should be created here and is None!")
+                if self.sub_process.stdout is not None:
+                    for raw_line in iter(self.sub_process.stdout.readline, b""):
+                        line = raw_line.decode(output_encoding, errors="backslashreplace").rstrip()
+                        self.log.info("%s", line)
+
+                self.sub_process.wait()
+                self.log.info("Command exited with return code %s", self.sub_process.returncode)
+
+                return_code: int = self.sub_process.returncode
```

Comment:
What I mean is this block is much much too big. A permission error can happen anywhere, and you may catch what you don’t want accidentally. A try block should be more localised and contain the lines you expect the exception to be raised from.

If there are multiple lines that can raise an exception scattered in wide range of lines (but not all lines in the range raise), it is preferred to do a quick check before any of the code runs and branch out depending on the check result.

Reply:
@uranusjr I updated to a simplified approach.